### PR TITLE
[stable/prometheus-pushgateway] Add persistence support

### PR DESCRIPTION
**Pull request by paprickar**
Signed-off-by: Patrick Kolze <patrick.kolze@doublecheck-it.de>

#### What this PR does / why we need it:
It allow the usage of persistent storage of the prometheus-pushgateway

#### Special notes for your reviewer:
Hey @gianrubio I hope this PR works for you - I tested it on GCP Kubernetes 1.13.6-gke.13 - on the bottom is the values file I used.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

My test values.yaml
```
# Default values for prometheus-pushgateway.
# This is a YAML-formatted file.
# Declare variables to be passed into your templates.
image:
  repository: prom/pushgateway
  tag: v0.9.0
  pullPolicy: IfNotPresent

service:
  type: ClusterIP
  port: 9091
  targetPort: 9091

# Optional pod annotations
podAnnotations: {}

# Optional pod labels
podLabels: {}

# Optional service labels
serviceLabels: {}

# Optional serviceAccount labels
serviceAccountLabels: {}

## Additional pushgateway container arguments
##
## for example: persistence.file: /data/pushgateway.data
extraArgs:
  persistence.file: /data/pushgateway.data
  persistence.interval: 1m
# Optional additional environment variables
extraVars: []

resources: {}
  # We usually recommend not to specify default resources and to leave this as a conscious
  # choice for the user. This also increases chances charts run on environments with little
  # resources, such as Minikube. If you do want to specify resources, uncomment the following
  # lines, adjust them as necessary, and remove the curly braces after 'resources:'.
  # limits:
  #   cpu: 200m
  #    memory: 50Mi
  # requests:
  #   cpu: 100m
  #   memory: 30Mi

serviceAccount:
  # Specifies whether a ServiceAccount should be created
  create: true
  # The name of the ServiceAccount to use.
  # If not set and create is true, a name is generated using the fullname template
  name:

## Configure ingress resource that allow you to access the
## pushgateway installation. Set up the URL
## ref: http://kubernetes.io/docs/user-guide/ingress/
##
ingress:
  ## Enable Ingress.
  ##
  enabled: false

    ## Annotations.
    ##
    # annotations:
    #   kubernetes.io/ingress.class: nginx
    #   kubernetes.io/tls-acme: 'true'

    ## Hostnames.
    ## Must be provided if Ingress is enabled.
    ##
    # hosts:
    #   - pushgateway.domain.com

    ## TLS configuration.
    ## Secrets must be manually created in the namespace.
    ##
    # tls:
    #   - secretName: pushgateway-tls
    #     hosts:
    #       - pushgateway.domain.com

tolerations: {}
  # - effect: NoSchedule
  #   operator: Exists

## Node labels for pushgateway pod assignment
## Ref: https://kubernetes.io/docs/user-guide/node-selection/
##
nodeSelector: {}

replicaCount: 1

## Security context to be added to push-gateway pods
##
securityContext:
  fsGroup: 65534
  runAsUser: 65534
  runAsNonRoot: true

## Affinity for pod assignment
## Ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity
affinity: {}

# Enable this if you're using https://github.com/coreos/prometheus-operator
serviceMonitor:
  enabled: true
  namespace: prometheus-pushgateway
  # fallback to the prometheus default unless specified
  interval: 10s
  ## Defaults to what's used if you follow CoreOS [Prometheus Install Instructions](https://github.com/helm/charts/tree/master/stable/prometheus-operator#tldr)
  ## [Prometheus Selector Label](https://github.com/helm/charts/tree/master/stable/prometheus-operator#prometheus-operator-1)
  ## [Kube Prometheus Selector Label](https://github.com/helm/charts/tree/master/stable/prometheus-operator#exporters)
  selector:
    prometheus: kube-prometheus
  # Retain the job and instance labels of the metrics pushed to the Pushgateway
  # [Scraping Pushgateway](https://github.com/prometheus/pushgateway#configure-the-pushgateway-as-a-target-to-scrape)
  honorLabels: true

# The values to set in the PodDisruptionBudget spec (minAvailable/maxUnavailable)
# If not set then a PodDisruptionBudget will not be created
podDisruptionBudget:

persistentVolume:
    ## If true, pushgateway will create/use a Persistent Volume Claim
    ## If false, use emptyDir
    ##
    enabled: true

    ## pushgateway data Persistent Volume access modes
    ## Must match those of existing PV or dynamic provisioner
    ## Ref: http://kubernetes.io/docs/user-guide/persistent-volumes/
    ##
    accessModes:
      - ReadWriteOnce

    ## pushgateway data Persistent Volume Claim annotations
    ##
    annotations: {}

    ## pushgateway data Persistent Volume existing claim name
    ## Requires pushgateway.persistentVolume.enabled: true
    ## If defined, PVC must be created manually before volume will be bound
    existingClaim: ""

    ## pushgateway data Persistent Volume mount root path
    ##
    mountPath: /data

    ## pushgateway data Persistent Volume size
    ##
    size: 2Gi

    ## pushgateway data Persistent Volume Storage Class
    ## If defined, storageClassName: <storageClass>
    ## If set to "-", storageClassName: "", which disables dynamic provisioning
    ## If undefined (the default) or set to null, no storageClassName spec is
    ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
    ##   GKE, AWS & OpenStack)
    ##
    # storageClass: "-"

    ## Subdirectory of pushgateway data Persistent Volume to mount
    ## Useful if the volume's root directory is not empty
    ##
    subPath: ""
```

### CODE REVIEW

**Comment by paprickar** [comment]
I hope the PR can be merged please remove stale flag

**Comment by paprickar** [comment]
Hey @gianrubio / @unguiculus,

could you please have a look into this PR. Please let me know if I should adapt something.

**Comment by pierluigilenoci** [comment]
When this PR will be merged make sense to redo the same also here: https://github.com/helm/charts/tree/master/stable/prometheus

File: stable/prometheus-pushgateway/Chart.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ apiVersion: v1
 appVersion: "0.9.0"
 description: A Helm chart for prometheus pushgateway
 name: prometheus-pushgateway
-version: 1.0.0
```
Comment (pierluigilenoci) [inline stable/prometheus-pushgateway/Chart.yaml]: You started from an old version of the master. You need to update your fork to be aligned before starting a PR.
https://github.com/helm/charts/blob/master/stable/prometheus-pushgateway/Chart.yaml

Comment (pierluigilenoci) [reply stable/prometheus-pushgateway/Chart.yaml]: The test failed for this reason. https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/helm_charts/16040/pull-charts-e2e/1179945867478568964/#1:build-log.txt%3A188

**Review by pierluigilenoci — CHANGES_REQUESTED** [review]

File: stable/prometheus-pushgateway/Chart.yaml (hunk 1 above)
Comment (pierluigilenoci) [reply stable/prometheus-pushgateway/Chart.yaml]: No. I was wrong. The PR remained unattended for too much time and then become old. The solution is the same. Redo the PR starting from an updated chart fork.

**Comment by pierluigilenoci** [comment]
@zanhsieh @gianrubio @unguiculus could you please give your "ok-to-test" approval on this PR?

**Comment by zanhsieh** [comment]
@pierluigilenoci 
It will take a long time to get this lgtm. According to the community member reported, @gianrubio , this chart owner, disappeared at least 17 days. You might find some [helm member](https://github.com/orgs/helm/people) to overwrite it, otherwise only the owner could merge. Sorry I am not the owner for this chart.

File: stable/prometheus-pushgateway/Chart.yaml (hunk 1 above)
Comment (paprickar) [reply stable/prometheus-pushgateway/Chart.yaml]: Thank you

**Comment by pierluigilenoci** [comment]
@gianrubio are you the owner?

**Comment by zanhsieh** [comment]
@pierluigilenoci 
Unfortunately some users report @gianrubio disappears for more than 17 days. You should ask some Helm member to overwrite the approval, and might self-promote the PR contributor to Helm member as this chart OWNERS after that.

**Comment by pierluigilenoci** [comment]
Done https://kubernetes.slack.com/archives/C0NH30761/p1571131197156800

File: stable/prometheus-pushgateway/templates/_helpers.tpl [hunk 2]
```diff
@@ -52,3 +52,26 @@ Create default labels
...
+
+{{- define "prometheus-pushgateway.matchLabels" -}}
+component: {{ .Values.name | quote }}
+{{ include "prometheus-pushgateway.common.matchLabels" . }}
+{{- end -}}
+
+{{- define "prometheus-pushgateway.common.metaLabels" -}}
+chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+heritage: {{ .Release.Service }}
+{{- end -}}
```
Comment (cstaud) [inline stable/prometheus-pushgateway/templates/_helpers.tpl]: @paprickar it seems the changes include more than just "Add persistence support". Pls open a seperate PR for different   functionality

**Review by cstaud — CHANGES_REQUESTED** [review]
Thank you for your work! Pls remove changes not required to fulfill the PRs intension. You also need to resolve the merge conflicts. Also I don't think the persistence should not be enabled by default.

**Comment by cstaud** [comment]
Also pls rename you PR according to the PR template. https://github.com/helm/charts/blob/master/.github/PULL_REQUEST_TEMPLATE.md

**Comment by jecnua** [comment]
Still valid. Please remove stale flag.

**Comment by pierluigilenoci** [comment]
@paprickar

**Comment by cstaud** [comment]
Due to no action from @paprickar staling this is the right way to go in my eyes.

**Comment by pierluigilenoci** [comment]
@cstaud I will push @paprickar phisically to fix this 😝

**Comment by cstaud** [comment]
However this is done: Pls rename the PR as advised above and resolve the current conflicts. Maybe it is easier to do this if you do a fresh checkout of the master and ship your changes within a new PR, but thats up to you :-)

File: stable/prometheus-pushgateway/templates/_helpers.tpl (hunk 2 above)
Comment (paprickar) [reply stable/prometheus-pushgateway/templates/_helpers.tpl]: @cstaud thanks for pointing this out.
I updated my PR and adapted to the current label handling. 
Let me know if this is correct.

**Comment by paprickar** [comment]
I was a bit more busy then expected lets see if I got it right now : >

**Comment by pierluigilenoci** [comment]
@cstaud the test was successfull. 🚀

**Comment by paprickar** [comment]
@cstaud could you have a look into it again? (it seems there was an update for the chart again so I updated it again - would be great to get this PR closed - if everything is fine)

File: stable/prometheus-pushgateway/templates/deployment.yaml [hunk 3]
```diff
@@ -23,13 +23,13 @@ spec:
         - name: pushgateway
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
           imagePullPolicy: {{ .Values.image.pullPolicy }}
+          args:
+          {{- range $key, $value := .Values.extraArgs }}
+            - --{{ $key }}={{ $value }}
```
Comment (davidkarlsen) [inline stable/prometheus-pushgateway/templates/deployment.yaml]: this breaks existing behaviour

Comment (paprickar) [reply stable/prometheus-pushgateway/templates/deployment.yaml]: Hey @davidkarlsen,

thanks for looking into it and finding this.

This time I installed the default prometheus-pushgateway with the default values as it is and then used my fork to upgrade with disks / persistence.

Here I installed the prometheus-pushgateway as it is
`helm3 install prometheus-pushgateway-test -f charts_upstream_master/stable/prometheus-pushgateway/values.yaml --namespace prometheus-pushgateway charts_upstream_master/stable/prometheus-pushgateway`

Here I used my fork with pvc support and my adjusted values.yaml
`helm3 upgrade prometheus-pushgateway-test -f prometheus-pushgateway-test/helm_chart_values-custom-chart.yaml --namespace prometheus-pushgateway charts/stable/prometheus-pushgateway`

I checked that I could reach the pushgateway and verified the configuration for persistence was changed / added.

This worked fine. If you have another test I should do please let me know.

**Comment by pierluigilenoci** [comment]
@cstaud @gianrubio could you please take a look?

**Comment by cstaud** [comment]
@pierluigilenoci @paprickar sry for the late reply there was some christmas/ new year stuff, you know ;-)

**Comment by pierluigilenoci** [comment]
@paprickar could you please bump again the chart version?

**Comment by paprickar** [comment]
Thank you @cstaud and @davidkarlsen for your help!

**Pull request merged**